DESIGN.md

```diff
@@ -0,0 +1,111 @@
+
+### SqlTable::Delete
+- Delete will be required to take in the schema version number that indicates the expected schema version number of the tuple being deleted
+
+### Catalog
+- The catalog will keep track of the visible schema versions for each transactions based on its timestamp. This schema version is then passed on to the SqlTable layer.
+
+### DataTable/Projected Row
+- DataTable::SelectIntoBuffer iterates across a projected_row/column’s column ids and fills in the data for each column
+
+    - For this operation we use the column_id VERSION_POINTER_COLUMN_ID as a sentinel id to represent a column that the DataTable should skip over and not fill in. We will go into detail as to why this happens within the architectural design below.
+
+
+## Architectural Design
+The design of this project will center around the modifications to SqlTable. The design for other components in the storage layer will remain unchanged. On a schema change the SqlTable will create a new DataTable that will store all the tuples inserted from that point on into the new version. To be lazy it will not modify already existing tuples to transform them into the latest version.  In order to support this lazy schema change we need two functionalities: maintaining tuples in multiple different schema versions and providing methods of transforming them into the desired version.
+
+### Multi-versioning
+To address the multi-versioning the SqlTable will maintain a map from the schema_version_number to a DataTable. There will be one DataTable for each schema version. The functionality for accessing tuples is already present in DataTable and this way SqlTable will only need to manage the two functionalities we described above. The rest of the functionalities will be handled by the already existing DataTable implementation. Furthermore, each block will maintain metadata of which version all of the tuples within the block belong to. Since each DataTable can only be a single version, a block cannot contains tuples from multiple versions.  Below is a description of the multiversion design for each of the SqlTable operations we are modifying and we will refer to expected version as the version of the tuple the user expects to see and actual version as the version the tuple is currently in.
+
+#### SqlTable::Insert
+Insert will always insert the passed in tuple into the DataTable of the schema version number passed in
+
+#### SqlTable::Update
+- Update has three cases
+    1. The expected schema version matches the actual schema version
+        - The update will happen in place on the DataTable of the actual schema version
+    2. The expected schema version doesn’t match the actual schema version but the update doesn’t touch any columns not in the actual schema version
+        - The update will happen in place on the DataTable of the expected schema version
```

Comment:
It is a bit confusing here. You have the following definition:
> Below we will refer to expected version as the version of the tuple the user expects to see and actual version as the version the tuple is currently in.

How can you update in place on the DataTable of expected schema version while the tuple is currently in a different version?

Reply:
Thanks!  This is a typo.  The in-place update occurs inside of the DataTable of the actual version.